runtime/io.js

```diff
@@ -139,7 +139,32 @@ function caml_ml_open_descriptor_in (fd)  {
   if(fd == 0 && fs_node_supported()){
     var fs = require('fs');
     refill = function () {
-      return caml_string_of_jsstring(fs.readFileSync(0, 'utf8'))};
+      var buf = globalThis.Buffer.alloc(256);
+      // Ref https://gist.github.com/espadrine/172658142820a356e1e0
+      var stdinFd;
+      var needsClose = false;
+      try {
+        stdinFd = fs.openSync('/dev/stdin', 'rs');
+        needsClose = true;
+      } catch (e) {
+        // Opening /dev/stdin can fail (on Windows or in pkg)
+        if (globalThis.process.platform === 'win32') {
+          // On Windows, we need to use the stdin fd
+          stdinFd = globalThis.process.stdin.fd;
+        } else {
+          // On Linux, we just want to use fd 0
+          // due to this nodejs bug: https://github.com/nodejs/node/issues/42826
+          stdinFd = 0;
+        }
+      }
+      var bytesRead = fs.readSync(fd, buf);
+      if (needsClose) {
+        fs.closeSync(stdinFd);
```

Comment:
Can't we leave the fd open for the next refill ?

Reply:
You don't want to leave nodejs file descriptors open for a long-lived period of time because they might get garbage collected. Currently, nodejs just logs a warning but they plan to throw in an upcoming version if a file descriptor would be GC'd

Reply:
How can they decide to GC a file descriptor ? Isn't it just an integer ? I find it weird to keep opening and closing stdin everytime you want to read from stdin. Can you point to some doc ?

Reply:
I assume that node maintains some resource pool underneath the fd integer. I don't have any docs to link, but recently encountered the deprecation message in a different project I work on.

Reply:
On further investigation, the error seems to be https://nodejs.org/api/deprecations.html#DEP0137
> Allowing a [fs.FileHandle](https://nodejs.org/api/fs.html#class-filehandle) object to be closed on garbage collection is deprecated. In the future, doing so might result in a thrown error that will terminate the process.
>
> Please ensure that all fs.FileHandle objects are explicitly closed using FileHandle.prototype.close() when the fs.FileHandle is no longer needed

Using the `fs.promises.open` API seems to return a different thing than `open` and `openSync` from the non-promise API (so confusing!)

So it is probably okay to open the fd once. However, where do you expect to close it? If it is open when the process tries to exit, won't the nodejs process be held open?

Reply:
> However, where do you expect to close it? If it is open when the process tries to exit, won't the nodejs process be held open?

I would close it whenever the use does `close_in stdin`

Reply:
@hhugo I'm confused then, do you want to move this "open file descriptor" logic to the place where you bind stdin for jsoo?

Reply:
I'll take a deeper look at this.

Reply:
https://github.com/ocsigen/js_of_ocaml/pull/1289 is refactoring the io logic in the runtime.
It adds a buffer to in_channel and makes reading from stdin similar to reading a regular file. Can you take a look at it, maybe test the PR and report back ?